### PR TITLE
*: add global variable tidb_schema_version_cache_limit to control infoschema cache size

### LINKED ISSUE #46524: stale read QPS is affected by DDL load
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Deploy a TiDB cluster.

2. run stale read load.
```
sysbench --config-file=sysbench.conf mussel_minimal.lua --threads=20 --read_staleness=-5 run
```

3. run DDL load by [loadgen](https://github.com/crazycs520/loadgen). Following load will use 1 thread to continuously create tables, from t1 to t100000.

```
loadgen bench --sql 'create table t#seq-val (id int key, b int)' --thread=1 --valmin=1 --valmax=100000 --port 23400 --host=192.168.78.2 --db=test2
```

### 2. What did you expect to see? (Required)

The stale read load QPS shouldn't be affected by irrelevant DDL load.

### 3. What did you see instead (Required)

The stale read load QPS is drop from 17200 to 250.

image

### 4. What is your TiDB version? (Required)

v6.5.4
master: 187573b0778fc757320052fcfb8fc403a01fd9cd

### 5. Analyze

I add some metrics in TiDB to debug: https://github.com/crazycs520/tidb/commit/06bf109f3be23152f9f4afca4aa5ac0222baf98f

image

The stale-read request is slow on the following execution stack

```
github.com/pingcap/tidb/domain.(*Domain).loadInfoSchema+0xdef						/data/cs/tidb/domain/domain.go:227
github.com/pingcap/tidb/domain.(*Domain).GetSnapshotInfoSchema+0x44					/data/cs/tidb/domain/domain.go:390
github.com/pingcap/tidb/sessiontxn/staleread.GetSessionSnapshotInfoSchema+0x53				/data/cs/tidb/sessiontxn/staleread/processor.go:320
github.com/pingcap/tidb/sessiontxn/staleread.(*staleReadProcessor).evaluateFromStmtTSOrSysVariable+0xdd	/data/cs/tidb/sessiontxn/staleread/processor.go:245
github.com/pingcap/tidb/sessiontxn/staleread.(*staleReadProcessor).OnSelectTable+0x17a			/data/cs/tidb/sessiontxn/staleread/processor.go:185
github.com/pingcap/tidb/planner/core.(*preprocessor).handleTableName+0xb6				/data/cs/tidb/planner/core/preprocess.go:1559
github.com/pingcap/tidb/planner/core.(*preprocessor).Leave+0x198					/data/cs/tidb/planner/core/preprocess.go:577
github.com/pingcap/tidb/parser/ast.(*TableName).Accept+0x164						/data/cs/tidb/parser/ast/dml.go:443
github.com/pingcap/tidb/parser/ast.(*TableSource).Accept+0x66						/data/cs/tidb/parser/ast/dml.go:599
github.com/pingcap/tidb/parser/ast.(*Join).Accept+0x66							/data/cs/tidb/parser/ast/dml.go:241
github.com/pingcap/tidb/parser/ast.(*TableRefsClause).Accept+0x64					/data/cs/tidb/parser/ast/dml.go:800
github.com/pingcap/tidb/parser/ast.(*SelectStmt).Accept+0x1fb						/data/cs/tidb/parser/ast/dml.go:1460
github.com/pingcap/tidb/parser/ast.(*TableSource).Accept+0x66						/data/cs/tidb/parser/ast/dml.go:599
github.com/pingcap/tidb/parser/ast.(*Join).Accept+0x66							/data/cs/tidb/parser/ast/dml.go:241
github.com/pingcap/tidb/parser/ast.(*TableRefsClause).Accept+0x64					/data/cs/tidb/parser/ast/dml.go:800
github.com/pingcap/tidb/parser/ast.(*SelectStmt).Accept+0x1fb						/data/cs/tidb/parser/ast/dml.go:1460
github.com/pingcap/tidb/planner/core.Preprocess+0x205							/data/cs/tidb/planner/core/preprocess.go:136
github.com/pingcap/tidb/executor.(*Compiler).Compile+0x3ed						/data/cs/tidb/executor/compiler.go:81
github.com/pingcap/tidb/session.(*session).ExecuteStmt+0x5d3						/data/cs/tidb/session/session.go:2214
github.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt+0xa6						/data/cs/tidb/server/driver_tidb.go:294
github.com/pingcap/tidb/server.(*clientConn).handleStmt+0x152						/data/cs/tidb/server/conn.go:2122
github.com/pingcap/tidb/server.(*clientConn).handleQuery+0x8f4						/data/cs/tidb/server/conn.go:1972
github.com/pingcap/tidb/server.(*clientConn).dispatch+0xf84						/data/cs/tidb/server/conn.go:1403
github.com/pingcap/tidb/server.(*clientConn).Run+0x232							/data/cs/tidb/server/conn.go:1152
github.com/pingcap/tidb/server.(*Server).onConn+0xa86							/data/cs/tidb/server/server.go:648
```

As you can see, the stale read query costs too much time to get snapshot infoschema.  This is the main cause of QPS degradation.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve?

Issue Number: close #46524

Problem Summary: add global variable tidb_schema_version_cache_limit to control infoschema cache size and add/refine metric.

### What is changed and how it works?

To fix #46524, we can simply enlarge the info-scheme cache capacity from 16 to 64. However, in scenarios with a large number of tables, this increases the risk of TiDB OOM.

So this PR just add global variable `tidb_schema_version_cache_limit ` to control infoschema cache size and refine metric in grafana to better diagnose this issue later. If you meet issue in #46524 with same root cause, you can enlarge the global variable `tidb_schema_version_cache_limit ` to fix it.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

1. Deploy a TiDB cluster.
2. run stale read load.
```
sysbench --config-file=sysbench.conf mussel_minimal.lua --threads=20 --read_staleness=-5 run
```

3. run DDL load by [loadgen](https://github.com/crazycs520/loadgen). Following load will use 1 thread to continuously create tables, from t1 to t100000.

```
loadgen bench --sql 'create table t#seq-val (id int key, b int)' --thread=1 --valmin=1 --valmax=100000 --db=test2
```
4. check the metrics.

image

- 17:12: Start running stale-read load, and the query QPS is 18000.
- 17:14: Start running DDL load, then the query QPS is dropping to 400. From the `Load Schema OPS` metrics, you can see many load snapshot schema OPS, this is what causes QPS to drop.
- 17:18: `set @@global. tidb_schema_version_cache_limit=64;`, then the query QPS return to 18000.

Side effects

- [ ] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- N/A

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by hawkingrei** [comment]
@crazycs520 Can you add the monitoring screenshots to PR description？

**Comment by crazycs520** [comment]
@xhebox @tangenta PTAL

File: domain/domain.go [hunk 1]
```diff
@@ -1046,7 +1046,7 @@ func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duratio
 		exit:                make(chan struct{}),
 		sysSessionPool:      newSessionPool(capacity, factory),
 		statsLease:          statsLease,
-		infoCache:           infoschema.NewCache(16),
```
Comment (xhebox) [inline domain/domain.go:1049]: I don't know if this is a good idea. Because caching cap could easily make OOM, too.

Comment (crazycs520) [reply domain/domain.go:1049]: I have tested it with 10000 tables, it doesn't consume much memory.

Comment (crazycs520) [reply domain/domain.go:1049]: How about add a global variable to control it?

Comment (zimulala) [reply domain/domain.go:1049]: Or test the 10,000,000 tables?

**Review by xhebox — APPROVED** [review]

File: domain/domain.go (hunk 1 above)
Comment (crazycs520) [reply domain/domain.go:1049]: Ok, I don't change the default value now, I add a global variable to control it.

**Comment by crazycs520** [comment]
@xhebox @zimulala @cfzjywxk PTAL

**Comment by cfzjywxk** [comment]
Need to discuss with @easonn7 first as there's session variable changes.

**Review by zimulala — APPROVED** [review]
LGTM

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**